### PR TITLE
More idiomatic for React, factor out functions

### PR DESCRIPTION
**Pull request by magopian**
This is following a quick review from @n1k0. Thanks for the advice, is there
anything else I should modify?

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -17,58 +39,36 @@ export class BMIComponent extends Component {
...
-      diagnostic = "obese";
-    }
+    var {height, weight, bmi} = calcBMI(this.state);
+    var [color, diagnostic] = getDiagnostic(bmi);
     return (
       <div>
         <div>
-          Height: {parseInt(height)}cm
-          <Slider id={"height"} value={height}
+          Height: {parseInt(height, 10)}cm
```
Comment (n1k0) [inline src/App.js]: It's a little weird to parse integers in the render part of the flow, especially as they're then rendered as string; I could totally see `calcBMI` returning properly typed values.

File: src/App.js [hunk 2]
```diff
@@ -10,6 +10,28 @@ class Slider extends Component {
   }
 }
 
+function calcBMI(data) {
```
Comment (n1k0) [inline src/App.js]: It's a little strange this accepts a `data` object, muted along internal processing here. This is especially annoying when you're passing a component state object; you should probably return an altered copy instead.

File: src/App.js [hunk 3]
```diff
@@ -17,58 +39,36 @@ export class BMIComponent extends Component {
...
   onSliderChange(fieldId, e) {
-    this.state[fieldId] = e.target.value;
+    var data = Object.assign({}, this.state);  // Copy the state.
+    data[fieldId] = e.target.value;
     if (fieldId != 'bmi') {
-      this.state.bmi = "";
+      data.bmi = "";
     }
-    this.setState(this.state);
+    this.setState(data);
```
Comment (n1k0) [inline src/App.js]: How about

```js
this.setState(Object.assign({}, this.state, {
  [fieldId]: e.target.value,
  bmi: fieldId !== "bmi" ? "" : this.state.bmi
});
```

File: src/App.js [hunk 4]
```diff
@@ -10,6 +10,28 @@ class Slider extends Component {
...
+function getDiagnostic(bmi) {
+    if (bmi < 18.5) {
+      return ["orange", "underweight"];
+    } else if (bmi < 25) {
+      return ["inherit", "normal"];
+    } else if (bmi < 30) {
+      return ["orange", "overweight"];
+    } else {
+      return ["red", "obese"];
+    }
```
Comment (n1k0) [inline src/App.js]: Nit: inconsistent indentation.

File: src/App.js [hunk 5]
```diff
@@ -17,58 +39,36 @@ export class BMIComponent extends Component {
   }
 
   onSliderChange(fieldId, e) {
-    this.state[fieldId] = e.target.value;
+    var data = Object.assign({}, this.state);  // Copy the state.
+    data[fieldId] = e.target.value;
```
Comment (n1k0) [inline src/App.js]: Maybe integer parsing could be done here, provided all received values are integers.

**Comment by n1k0** [comment]
Nice :) Some nits and minor suggestions you might want to address.

File: src/App.js (hunk 3 above)
Comment (magopian) [inline src/App.js]: won't work if `fieldId` is `bmi`. Also, it's not very readable imho

File: src/App.js (hunk 5 above)
Comment (magopian) [inline src/App.js]: The values are then computed in the `calcBMI` function, which might return floats. I've moved the calls to `parseInt` there

**Comment by magopian** [comment]
Thanks for the reviews! I've updated the code, what do you think?

File: src/App.js [hunk 6]
```diff
@@ -17,58 +40,37 @@ export class BMIComponent extends Component {
...
-    } else if (bmi < 30) {
-      color = "orange";
-      diagnostic = "overweight";
-    } else {
-      color = "red";
-      diagnostic = "obese";
-    }
+    var [height, weight, bmi] = calcBMI(this.state.height, this.state.weight,
+                                        this.state.bmi);
+    var [color, diagnostic] = getDiagnostic(bmi);
```
Comment (n1k0) [inline src/App.js]: Note that you can replace `var` with `const`, here and elsewhere. That won't (sadly) turn these into immutable structures, but at least that prevents their value to be reassigned.

File: src/App.js [hunk 7]
```diff
@@ -10,6 +10,29 @@ class Slider extends Component {
...
 
+function calcBMI(height, weight, bmi) {
+  var h = height / 100;  // Height in meters instead of centimeters;
+  if (!bmi) {
+    bmi = weight / (h * h);  // Set the BMI from the weight.
+  } else {
+    weight = bmi * h * h;  // Set the weight from the BMI.
+  }
+  var result = [height, weight, bmi];
+  return result.map(x => parseInt(x, 10));
```
Comment (n1k0) [inline src/App.js]: Nice :)

**Comment by n1k0** [comment]
r+ :dancers:

**Comment by magopian** [comment]
Updated, thanks again!

**Pull request merged**